core/examples/http_bench_json_ops.rs

```diff
@@ -78,19 +79,33 @@ struct TcpStream {
 }
 
 impl TcpStream {
-  async fn read(self: Rc<Self>, buf: &mut [u8]) -> Result<usize, Error> {
+  async fn read(
+    self: Rc<Self>,
+    mut buf: ZeroCopyBuf,
+  ) -> Result<usize, AnyError> {
     let mut rd = RcRef::map(&self, |r| &r.rd).borrow_mut().await;
     let cancel = RcRef::map(self, |r| &r.cancel);
-    rd.read(buf).try_or_cancel(cancel).await
+    rd.read(&mut buf)
+      .try_or_cancel(cancel)
+      .await
+      .map_err(AnyError::from)
   }
 
-  async fn write(self: Rc<Self>, buf: &[u8]) -> Result<usize, Error> {
+  async fn write(self: Rc<Self>, buf: ZeroCopyBuf) -> Result<usize, AnyError> {
     let mut wr = RcRef::map(self, |r| &r.wr).borrow_mut().await;
-    wr.write(buf).await
+    wr.write(&buf).await.map_err(AnyError::from)
   }
 }
 
 impl Resource for TcpStream {
+  fn read(self: Rc<Self>, buf: ZeroCopyBuf) -> Option<AsyncResult<usize>> {
```

Comment:
I think I would prefer the return type to be just AsyncResult.
The default implementation can return some sort of "not supported" error.

Reply:
One advantage of the `Option<AsyncResult<...>>` approach is that when shutting down a parent resource we can `filter_map` its children and shutdown that collection error-ing in `shutdown()` would make that trickier or require a "can shutdown" tag

Reply:
You can also make the default implementation return Ok() if thats what you're after.
That might be ok (pun not intended) in the specific case of shutdown - I have to think about that.

But I think we should discuss use cases and potential merit of parent/child resources first.
Do you have an example in mind of a resource that would benefit from having child resources?

Reply:
Http is the main example, we currently track child resources JS-side. Http has a rank 3 hierarchy (listener > conn > req/resp), other listeners are rank 2.

That would work, it would just be theoretically less efficient

Reply:
> Http is the main example, we currently track child resources JS-side.

I am no longer convinced HTTP really needs child resources though. Potentially between the TCP listener and the HttpConn, but I suspect you were thinking of HttpConn => HttpStream as parent-child.

The filter_map shutdown behavior isn't appropriate for HttpCon resources; `shutdown` has a specific meaning - it tells the receiving end of a connection that the sender has sent all the data it had intended to send (rather than e.g. the connection was closed due to a crash or connection error). HttpConn, which behaves like a listening socket rather than a data stream, doesn't have a shutdown method.

We could consider expanding the meaning of `shutdown()` to mean `close gracefully` in a more general way.
In that case `HttpConn` could have a shutdown method, which would make it stop accepting new HTTP requests while letting already-accepted ones finish up -- however that would not involve calling shutdown on its child resources.

Reply:
I'll drop the option, it's not difficult to change this if needed down the line

Reply:
Actually I would rather keep the Option for now change it if necessary in a follow-up PR.

It could be more ergonomic to drop the Option if other ops take a readable resource and are doing their own `.read()` calls on them so it's an at runtime `AsyncRead` of sorts